File: Java/828.java
__________________________________________________________________________________________________
sample 2 ms submission
class Solution {
    public int uniqueLetterString(String S) {
        int[] lastPosition = new int[26];
        int[] contribution = new int[26];
        int result = 0;
        
        // at each iteration, we count the contribution of all the characters to all the substrings ending till that point
        for(int i = 0; i < S.length(); i++) {
            int current = S.charAt(i) - 'A';
            
            // update contribution to current character.
            // total # of substrings ending at i are i + 1. If it was a unique character, it would contribute to all of the substrings and its contribution is i + 1
            // if it is repeating, then it was contributed previously, so we will subtract its previous contribution
            // these are the contributions for strings which start after this character's last occurence and end at i
            int totalNumOfSubstrings = i + 1;
            contribution[current] = totalNumOfSubstrings - lastPosition[current];
            
            // the contribution of all other charcters will remain the same
            
            // count the current answer by adding all the contributions, only 26 letters in alphabet
            int curr = 0;
            for(int j = 0; j < 26; j++) {
                curr += contribution[j];
            }
            
            // add current value to final answer
            result += curr;
            
            // update last position of the current character
            lastPosition[current] = i + 1;
        }
        return result;
    }
}
__________________________________________________________________________________________________
sample 35660 kb submission
class Solution {
     public int uniqueLetterString(String S) {
        Map<Character, List<Integer>> index = new HashMap();
        for (int i = 0; i < S.length(); ++i) {
            char c = S.charAt(i);
            index.computeIfAbsent(c, x-> new ArrayList<Integer>()).add(i);
        }

        long ans = 0;
        for (List<Integer> A: index.values()) {
            for (int i = 0; i < A.size(); ++i) {
                long prev = i > 0 ? A.get(i-1) : -1;
                long next = i < A.size() - 1 ? A.get(i+1) : S.length();
                ans += (A.get(i) - prev) * (next - A.get(i));
            }
        }

        return (int) ans % 1_000_000_007;
    }
}
__________________________________________________________________________________________________

File: Python3/828.py
__________________________________________________________________________________________________
sample 60 ms submission
import string


class Solution:
    """O(n) solution

    Think about string "XAXAXXAX" and focus on making the second "A" a unique
    character.
    We can take "XA(XAXX)AX" and between "()" is our substring. To make the
    second "A" counted as a uniq character, we need to:

    1. insert "(" somewhere between the first and second A
    2. insert ")" somewhere between the second and third A

    For step 1 we have "A(XA" and "AX(A", 2 possibility.
    For step 2 we have "A)XXA", "AX)XA" and "AXX)A", 3 possibilities.

    So there are in total 2 * 3 = 6 ways to make the second A a unique character
    in a substring.
    In other words, there are only 6 substring, in which this A contribute 1
    point as unique string.

    Instead of counting all unique characters and struggling with all possible
    substrings, we can count for every char in S, how many ways to be found as a
    unique char. We count and sum, and it will be the answer.
    """

    MODULO_BASE = 10 ** 9 + 7

    def uniqueLetterString(self, s: str) -> int:
        # index[c] stores the last two indexes of appearance of c.
        index = {c: [-1, -1] for c in string.ascii_uppercase}
        total = 0
        for i, c in enumerate(s):
            k, j = index[c]
            total += (i - j) * (j - k)
            index[c] = [j, i]
        # Calculate the last occurrence for each letter as well.
        for c in index:
            k, j = index[c]
            total += (len(s) - j) * (j - k)
        return total % self.MODULO_BASE
__________________________________________________________________________________________________
sample 68 ms submission
class Solution:
    def uniqueLetterString(self, S: str) -> int:
        # https://leetcode.com/problems/unique-letter-string/discuss/128952/One-pass-O(N)-Straight-Forward
        # https://www.icode9.com/content-4-25511.html
        res = 0
        kMod = 1e9 + 7
        char_pos = collections.defaultdict(lambda : [-1, -1])
        for i, c in enumerate(S): # process all unique but last one
            k, j = char_pos[c]
            res += (j - k) * (i - j)
            char_pos[c] = [j, i]
        n = len(S)
        for  c in char_pos: # ATTN only process char_pos.keys(if you process S again,you will duplicate if some appear more than once) only process process last unique
            k, j = char_pos[c]
            res += (n - j) * (j - k)
        
        return int(res % kMod)
__________________________________________________________________________________________________
